proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java

```diff
@@ -596,6 +599,20 @@ public void handle(Commands commands) throws Exception
         }
     }
 
+    @Override
+    public void handle(Chat chat) throws Exception
+    {
+        ServerChatEvent chatEvent = new ServerChatEvent( con.getServer(), con, chat.getMessage(), chat.getPosition() );
+        if ( !bungee.getPluginManager().callEvent( chatEvent ).isCancelled() )
+        {
+            // Be aware that because of this, all server-sent chat messages are parsed and re-serialised
+            chat.setMessage( ComponentSerializer.toString(chatEvent.getMessage()) );
```

Comment:
Maybe it would be worth add a field that keeps track if the message was modifyed. When you could determine if (re)serialization would be needed.

Reply:
Oh lol that's what I just said above.
Probably worth trying to see if it's possible

Reply:
I would do it like this

```diff
diff --git a/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java b/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java
index b9f54989..51c80b36 100644
--- a/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java
+++ b/api/src/main/java/net/md_5/bungee/api/event/ServerChatEvent.java
@@ -1,7 +1,8 @@
 package net.md_5.bungee.api.event;
 
-import lombok.Data;
 import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
 import lombok.ToString;
 import net.md_5.bungee.api.ChatMessageType;
 import net.md_5.bungee.api.chat.BaseComponent;
@@ -12,7 +13,7 @@ import net.md_5.bungee.chat.ComponentSerializer;
 /**
  * Event called when a server sends a message to a player.
  */
-@Data
+@Getter
 @ToString(callSuper = true)
 @EqualsAndHashCode(callSuper = true)
 public class ServerChatEvent extends TargetedEvent implements Cancellable
@@ -21,6 +22,7 @@ public class ServerChatEvent extends TargetedEvent implements Cancellable
     /**
      * Cancelled state.
      */
+    @Setter
     private boolean cancelled;
     /**
      * Text contained in this chat.
@@ -29,7 +31,12 @@ public class ServerChatEvent extends TargetedEvent implements Cancellable
     /**
      * Type of message/display location
      */
+    @Setter
     private ChatMessageType position;
+    /**
+     * Interal state flag to determine if the message filed has changed.
+     */
+    private boolean messageWasEdited = false;
 
     public ServerChatEvent(Connection sender, Connection receiver, String json, byte position)
     {
@@ -37,4 +44,9 @@ public class ServerChatEvent extends TargetedEvent implements Cancellable
         this.message = ComponentSerializer.parse( json );
         this.position = ChatMessageType.values()[position];
     }
+
+    public void setMessage(BaseComponent[] message) {
+        this.messageWasEdited = true;
+        this.message = message;
+    }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 4e992012..adadb411 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -606,7 +606,9 @@ public class DownstreamBridge extends PacketHandler
         if ( !bungee.getPluginManager().callEvent( chatEvent ).isCancelled() )
         {
             // Be aware that because of this, all server-sent chat messages are parsed and re-serialised
-            chat.setMessage( ComponentSerializer.toString(chatEvent.getMessage()) );
+            if (chatEvent.isMessageWasEdited()) {
+                chat.setMessage( ComponentSerializer.toString(chatEvent.getMessage()) );
+            }
             chat.setPosition( ( byte ) chatEvent.getPosition().ordinal() );
             con.unsafe().sendPacket( chat );
         }

```